hathor/websocket/protocol.py

```diff
@@ -30,23 +42,230 @@ class HathorAdminWebsocketProtocol(WebSocketServerProtocol):
+                    'type': 'stream:history:error',
+                    'id': stream_id,
+                    'errmsg': 'Streaming is already opened.'
+                }))
+                return
+
+            assert self._manual_address_iter is not None
+
+            try:
+                self._manual_address_iter.add_addresses(addresses, last)
+            except LimitExceeded:
+                self.sendMessage(json_dumpb({
+                    'type': 'stream:history:error',
+                    'id': stream_id,
+                    'errmsg': 'List of addresses is too long.'
+                }))
+                return
+            except InvalidAddress as exc:
+                self.sendMessage(json_dumpb({
+                    'type': 'stream:history:error',
+                    'id': stream_id,
+                    'errmsg': f'Invalid address: {exc}'
+                }))
+                return
+
+            self.log.info('Adding addresses to a websocket manual streaming',
+                          stream_id=stream_id,
+                          addresses=addresses,
+                          last=last)
+            return
+
+        gap_limit = message.get('gap-limit', 20)
+        if gap_limit > self.MAX_GAP_LIMIT:
+            self.sendMessage(json_dumpb({
+                'type': 'stream:history:error',
+                'id': stream_id,
+                'errmsg': f'GAP limit is too big. Maximum: {self.MAX_GAP_LIMIT}'
+            }))
+            return
+
+        if not first:
+            self.sendMessage(json_dumpb({
+                'type': 'stream:history:error',
+                'id': stream_id,
+                'errmsg': 'Streaming not found. You must send first=true in your first message.'
+            }))
+            return
+
+        address_iter = ManualAddressSequencer()
+        try:
+            address_iter.add_addresses(addresses, last)
+        except LimitExceeded:
+            self.sendMessage(json_dumpb({
+                'type': 'stream:history:error',
+                'id': stream_id,
+                'errmsg': 'List of pending addresses is too long.'
+            }))
+            return
+        except InvalidAddress as exc:
+            self.sendMessage(json_dumpb({
+                'type': 'stream:history:error',
+                'id': stream_id,
+                'errmsg': f'Invalid address: {exc}'
+            }))
+            return
+
+        search = GAPLimitSearch(self.factory.manager, address_iter, gap_limit)
+        self._manual_address_iter = address_iter
+        self._history_streamer = HistoryStreamer(protocol=self, stream_id=stream_id, search=search)
+        deferred = self._history_streamer.start()
+        deferred.addBoth(self._streamer_callback)
+        self.log.info('opening a websocket manual streaming',
+                      stream_id=stream_id,
+                      addresses=addresses,
+                      gap_limit=gap_limit,
+                      last=last)
```

Comment:
I can't find the other code. Is this the xpub log message?

Reply:
No, it's in the end of the `_open_history_xpub_streamer()` method. Both contain this part:

```python
search = GAPLimitSearch(self.factory.manager, address_iter, gap_limit)
self._history_streamer = HistoryStreamer(protocol=self, stream_id=stream_id, search=search)
deferred = self._history_streamer.start()
deferred.addBoth(self._streamer_callback)
```

Since there's common deferred logic, I think this should be in its own method.

Reply:
Got it. You're right. Done! Thanks!

Reply:
Great, thanks!